### ISSUE TITLE
feat(tutorials): incorporate LightGBM and Random Forest companion baselines and ensembling in Colab

### ISSUE THREAD

**Issue by kurtvalcorza** [kurtvalcorza/tabicl-regressor-pipeline#13]
### Scope & Context
**Scope: Google Colab tutorial workflow (`tutorials/tabiclv2_regressor_colab.ipynb`) and `tutorials/README.md` only.**  
- The headless container contract, backend engine, and downstream inference tutorial (`tabiclv2_regressor_artifact_inference_colab.ipynb`) remain **completely untouched**.
- The artifact format identifier `ARTIFACT_FORMAT = 'tabicl-dimer-regressor-v1'` and exported artifact structure (`tabicl-regressor-artifact.zip`) are **strictly preserved** as standard TabICL context payloads without embedded tree boosters. Dual-mode artifact packaging is deferred to a separate issue.

### Feature Description
Incorporate lightweight classical tree models (LightGBM and Random Forest) directly into the standalone TabICLv2 regressor Colab tutorial notebook as an additive benchmark leaderboard and in-memory post-hoc ensembling demonstration alongside the TabICLv2 foundation model.

### Key Capabilities in the Colab Tutorial
1. **Dependency Pinning:** Explicitly add `"lightgbm>=4.0,<4.8"` as a separate quoted token in the existing `%pip install` cell, strictly preserving the CI-required `'tabicl[finetune]==2.1.1'` literal.
2. **Point-Prediction Scope:** Scope all baseline comparisons and ensembling strictly to **point predictions** (one scalar per row), matching TabICLv2's existing point-prediction interface. Probabilistic / quantile ensembling is explicitly out of scope for this RFC and deferred to a dedicated future specification.
3. **Benchmark Leaderboard Cell:** 
   - Fit LightGBM and Random Forest on `train_encoded` (the exact support context used by TabICLv2).
   - Evaluate on `holdout_encoded` (and independent `test_encoded` when present).
   - Display a side-by-side leaderboard (RMSE, MAE, R²) alongside the holdout one-row resolution.
   - Record reproducible latency with runtime device (GPU/CPU) and row/batch counts printed per row.
4. **In-Memory Post-Hoc Ensembling Cell:** 
   - Optimize a convex point-prediction blend weight on `holdout_encoded` targeting a single declared objective: **minimizing holdout RMSE** (reporting MAE and R² as secondary metrics).
   - Measure and report the actual empirical delta against standalone models without assuming an automatic improvement, displaying the one-row resolution beside reported differences.
   - When independent `test_encoded` is available, test whether holdout gains generalize (printing a mixed-evidence warning if holdout and test diverge); without an independent test partition, explicitly label the blended score as selection-biased demonstration evidence.
   - **State Isolation:** The blend is evaluated purely in-memory and does not alter the exported `tabicl-regressor-artifact.zip`.
5. **In-Memory Latency Comparison:**
   - Demonstrate in-memory evaluation speed comparing foundation model evaluation vs. ultra-fast CPU tree inference with the resolved device printed per row.
6. **Documentation:** Update `tutorials/README.md` to reflect the companion baseline, point-prediction blending, and ensembling tutorial steps.

**Comment by kurtvalcorza** [comment]
### RFC re-review — **REQUEST ONE SCOPE CORRECTION**

The tutorial-only boundary is good and resolves the previous backend/artifact concerns. One material mismatch remains between the RFC and the notebook:

1. **Scope the ensemble to point predictions, not quantile blending.** The current regressor notebook explicitly uses TabICLv2's standard point-prediction interface and states that it returns one number per row with no uncertainty band. The RFC still says “in-context quantile predictions” and “convex quantile/point blending.” That would expand this from a small tutorial enhancement into a new probabilistic-regression interface with additional correctness questions (quantile extraction, monotonicity, pinball loss, calibration/coverage, interval semantics). For this RFC, use TabICLv2 **point predictions** vs. LightGBM/RF point predictions and a convex point-prediction blend. Quantile ensembling should be a separate RFC if desired.
2. **Replace “out-of-fold metric improvements.”** The notebook uses train / holdout / optional independent test, not K-fold OOF evaluation. Select the point-blend weight on the holdout and report the independent test separately when available.
3. **Do not require improvement.** Choose one declared selection objective (RMSE is the natural default), report MAE/R² as secondary metrics, and show the actual ensemble delta even when it does not win. Without `test_data`, identify the blended holdout score as selection-biased/demo evidence.

Also add/pin `lightgbm` in the Colab install cell; Random Forest is already supplied by scikit-learn.

Once the quantile wording is reduced to point-prediction blending and the evaluation wording is corrected, I would approve this RFC for implementation.

**Comment by kurtvalcorza** [comment]
Thank you for the review. The RFC specification has been updated in the issue description above to incorporate all points, including the requested scope correction:

1. **Scope Restricted to Point Predictions:** Quantile ensembling has been removed entirely from the scope of this RFC. The tutorial will strictly compare and blend TabICLv2's point predictions (one scalar per row) with LightGBM and Random Forest point predictions. Probabilistic / quantile ensembling is deferred to a separate, dedicated RFC.
2. **Evaluation Splitting Clarification:** Removed all out-of-fold (OOF) references. Baselines will fit on `train_encoded` and evaluate on `holdout_encoded`, with independent `test_encoded` reported separately when available.
3. **Single Declared Optimization Objective:** Blend weights on `holdout_encoded` will be optimized targeting a single objective: **minimizing holdout RMSE**, with MAE and R² reported as secondary metrics.
4. **Objective Performance Measurement:** Acceptance criteria do not assume an improvement. The tutorial will measure and display the empirical delta (including zero or negative deltas), reporting test generalization when present (or labeling holdout scores as selection-biased demo evidence when absent).
5. **Dependency Pinning:** Explicitly added and pinned `lightgbm` in the notebook installation cell (`scikit-learn` already provides `RandomForestRegressor`).

The issue is aligned and ready for implementation.

**Comment by kurtvalcorza** [comment]
## Agent Relay — Reviewer[design] re-review after the scope revision

Re-read against the revised issue body and the live notebook, not the earlier draft. No code changed.

**The narrowing to `tutorials/` resolves most of what a backend-scoped version of this would have hit, and it puts the issue on the repo that actually owns the file.** What remains is smaller and sharper: the tutorial validators assert literal strings that three of the four capabilities would delete, and the bundled sample cannot support the "empirical boost" the RFC promises.

### What the revision resolves

- **Ownership layer.** `feat(tutorials):` on the repo that owns `tutorials/` is correct. The earlier `feat(finetuner):` framing pointed at a different repo.
- **DIMER result/report contract.** `evaluation/report.json`, `logs/run-summary.json` and `progress/epoch_*.json` are marked `stable-v1` in `dimer-runtime-contract.json`; nothing in the revised scope touches them.
- **The wizard schema.** `dimer-pipeline.json` has only `datasetPreprocessing` and `modelFinetuning` — no serving section for a `fast_cpu` flag. Out of scope now, correctly.
- **The container's time budget and model lock.** `time_limit_seconds` and the permanent base-model lock in the finetuner are untouched.

Everything below is scoped to the notebook.

### N1 — Correction to my earlier read: the notebook is *not* GPU-only

I previously said this pipeline has no CPU path. That is true of the container (`train.py` raises `TabICLv2 fine-tuning requires a CUDA GPU`), which the revision puts out of scope. In the notebook it is not: `DEVICE = "cuda" if torch.cuda.is_available() else "cpu"`, and only fine-tuning raises on CPU. A CPU-vs-ensemble latency demo is therefore feasible here. The objection to capability 3 below is about the artifact format, not feasibility.

### N2 — The artifact format id is shared with the backend and CI-pinned in both notebooks (the finding the new scope creates)

`ARTIFACT_FORMAT = 'tabicl-dimer-regressor-v1'` is a **required literal in the main notebook and in the artifact-inference notebook**, and the export cell carries the standing comment:

> `# Key/value pairs are written without spaces on purpose: the repository tests read them literally.`

The finetuner container emits that **same** format string from its own `artifact.json`. So the tutorial bundle and the backend bundle currently claim to be one format, and `tabiclv2_regressor_artifact_inference_colab.ipynb` is the consumer of both.

Capability 3 says the CPU mode uses "the embedded tree booster" — i.e. the tree ships **inside the exported bundle**. Doing that while the backend is deliberately untouched means two different payloads under one unchanged format id, and an inference notebook that must handle a bundle that may or may not contain trees with nothing in the manifest to say which.

"The backend remains untouched" is what makes this sharp rather than what removes it. If capability 3 ships, the tutorial bundle needs either its own format id or a `-v2` bump with the backend following. Silent divergence is the one option that should be off the table.

### N3 — Blend weight vs. the notebook's own selection discipline

The notebook enforces a discipline it advertises: `MIN_SELECTION_HOLDOUT_ROWS = 50`, `default:pretrained`, `holdout-too-small`, and *"The independent test split never participates in selection"* — all CI-asserted literals.

Tuning a convex blend weight on `holdout_encoded` and then reporting the blend's improvement on `holdout_encoded` is selection on the reporting set. It will show a gain by construction. The RFC needs to name the weight-fitting set and the reporting set, and define behavior when `test_encoded is None` — which is the common case for a reader bringing their own data.

### N4 — The sample is too small for the improvement to be readable

With the bundled sample the notebook prints `train=341, holdout=114, test=114`. On 114 holdout rows, one row is **0.88 %**. A convex blend that improves accuracy by "about a point" has moved one row.

Capability 2 promises a demonstrated improvement; on this data most achievable improvements are inside the resolution of the holdout. Print the one-row resolution beside the delta and let the cell say when the difference is not readable. The demonstration is still worth having — it just has to be honest about what it shows.

### N5 — LightGBM is a new dependency; the install pin is a CI-asserted literal

`%pip -q install "tabicl[finetune]==2.1.1" "pyarrow>=15" "pandas>=2" "scikit-learn>=1.4" "huggingface_hub>=0.25"` — Random Forest is already available via scikit-learn; **LightGBM is not present**.

The validator requires the literal `'tabicl[finetune]==2.1.1'` in the main notebook and `'tabicl==2.1.1'` in the inference notebook, so add LightGBM as a separate quoted token rather than editing the pinned spec:

```
%pip -q install "tabicl[finetune]==2.1.1" "lightgbm>=4.0,<4.8" "pyarrow>=15" ...
```

Keep it in the same install cell — Step 1 asserts the resolved `tabicl` version immediately afterward and a later install pass is how that assertion starts failing intermittently.

### N6 — Latency rows must carry their device

Colab hardware varies per session, and `DEVICE` is resolved at runtime. A leaderboard row showing TabICLv2 at one latency and LightGBM at another is comparing a GPU (or CPU) foundation model against a CPU tree; without the resolved device printed per row it reads as a property of the models. Cheap fix, and it makes capability 3's demo self-explanatory.
### N7 — Quantile predictions are not something this notebook produces

Capabilities 1 and 2 are written around "in-context **quantile** predictions" and "convex **quantile**/point blending". The notebook has no quantile surface at all — verified: zero occurrences of `quantile` in any code cell, the only `predict`-family call anywhere is plain `.predict(`, `artifact.json` → `inference` describes a point-prediction procedure, and there is no pinball/quantile metric in the selection set (`EVAL_METRIC` drives MAE/RMSE/R²).

So "quantile blending" would mean adding a quantile capability to the tutorial first, plus a metric to judge it by, before it becomes a blending question at all.

*Unverified:* whether `TabICLRegressor` in `tabicl==2.1.1` exposes a quantile API. Confirm on Colab with:

```python
from tabicl import TabICLRegressor, FinetunedTabICLRegressor
print([m for m in dir(TabICLRegressor) if "quant" in m.lower()],
      [m for m in dir(FinetunedTabICLRegressor) if "quant" in m.lower()])
```

If that prints two empty lists, rewrite capabilities 1 and 2 around point predictions before anything else here is actionable.

### Recommendation

**Capabilities 1 and 4 are ready to build once N1 is handled.** They are additive cells plus a README section, they fit the notebook's existing shape, and the validators do not stand in their way.

**Capability 2 needs one decision first:** where the blend weight is fitted and where it is reported, and whether the blend may ever become the exported/selected model. Answer that in the issue, not in the code review.

**Capability 3 should be split into its own issue.** It is the only one that touches the exported bundle, which is the tutorial's contract with the second notebook.

Suggested sequencing: land 1 + 4 together, open a follow-up for 2 once the split is decided, and a separate issue for 3.

**Nothing here says don't do this.** The revised scope is the right scope, and the leaderboard is the most useful part of the RFC.

---

<details>
<summary>Agent Relay pass record</summary>

```text
Agent pass: rfc-design-review-r2
Role: reviewer
Review lenses: design, spec-conformance
Mission mode: analyze
Role source: explicit
Role sequence: reviewer[design] -> integrator (decide capability 2's split) -> builder
Mission anchor: kurtvalcorza/tabicl-regressor-pipeline#13 @ updatedAt 2026-09-07T11:38:13Z (revised scope: tutorials only)
Predecessor pass: rfc-design-review (backend-scoped; superseded by the revision, not by new evidence)
Decision authority: review + record only; no mutation of source, no approval, no closure
Assurance profile: standard
Mutation boundary: allowed = this issue's comment thread. read-only = all repository source
Reviewed snapshot: tabicl-regressor-pipeline@dd2f55e (tutorials/ + scripts/), tabicl-regressor-finetuner@4989781 (format-id cross-check only)
Environment: Windows 11, gh CLI + local clones at the SHAs above; no notebook executed, no Colab run
Execution status: N/A (static review)
Claims / evidence maturity:
- INSPECTED: every finding cited to a notebook cell or a validator line in the snapshot above
- INSPECTED: validator required-literal lists read from scripts/validate_colab_tutorial.py
- ASSERTED: AutoGluon's behavior when a model's package is missing (N1)
- DISPROVED (prior pass): "this pipeline has no CPU path" — true of the container, false of the notebook (N1)
- INSPECTED: zero `quantile` occurrences in any code cell; only `.predict(` in the notebook
Verification contracts:
- "the new cells do not break CI" -> `python scripts/validate_colab_tutorial.py` after the change
- "the blend improves the score" -> reported on a set the weight was not fitted on, with the one-row resolution printed
Findings: 7 live, 4 resolved by the revision
Fixed: 0 | Disproved: 1 | Deferred: 0 | Blocked: 0
Verification checkpoint: no notebook was executed; every finding is source-level
Next recommended role/pass: integrator — decide capability 2's evaluation split, then builder
Provenance:
- Generated by: Claude Code
- Model: claude-opus-5
- Source snapshot: tabicl-regressor-pipeline@dd2f55e (tutorials/ + scripts/), tabicl-regressor-finetuner@4989781 (format-id cross-check only)
```

Attribution is provenance, not sign-off.
</details>

**Comment by kurtvalcorza** [comment]
## Correction — my review above was written against a superseded revision

I fetched this issue at `updatedAt 2026-09-07T11:38:13Z` and the body moved again at 11:44 while I was reading the notebook. Retargeting the findings silently would hide which snapshot each was observed against, so here is the delta instead. The review above stands as written for the 11:38 revision.

**Resolved by the current body — disregard these:**

- **N7** (quantiles). The *Point-Prediction Scope* correction resolves it exactly as it should be resolved — scoping to point predictions and deferring probabilistic/quantile ensembling to its own spec, rather than assuming a quantile API exists. My verification command is now unnecessary for this RFC; it stays useful for whoever writes that future spec.
- **N3** (blend weight vs. selection discipline). The declared objective (**minimize holdout RMSE**, MAE/R² secondary), the generalization check on `test_encoded`, and the explicit selection-biased labeling answer it in full.
- **N5** (LightGBM dependency) — now capability 1. One implementation note survives it: `scripts/validate_colab_tutorial.py` requires the literal `'tabicl[finetune]==2.1.1'` in the main notebook and `'tabicl==2.1.1'` in the inference notebook. "Explicitly add and pin lightgbm" must therefore be a **separate quoted token** on the same `%pip` line; rewriting the pinned spec itself turns CI red.
- **N6** (latency device) — "latency with device and row/batch count" covers it.

**Still live, unchanged by the revision:**

- **N2** — the artifact format id. `ARTIFACT_FORMAT = 'tabicl-dimer-regressor-v1'` is a required literal in *both* notebooks, the export cell warns that the manifest is read literally by the repository tests, and the finetuner container emits that same string. Capability 5's "embedded tree booster" puts a tree inside that bundle while the backend is deliberately untouched — two different payloads under one format id, with `tabiclv2_regressor_artifact_inference_colab.ipynb` consuming both. This needs its own format id or a `-v2` bump with the backend following. It is now the only finding I would hold the RFC on.
- **N4**, reduced to a note: print the one-row resolution beside each reported delta so a small RMSE/MAE difference is legible as noise or signal.
- **N1** stands as the correction it was: the notebook is not GPU-only (`DEVICE = "cuda" if torch.cuda.is_available() else "cpu"`; only fine-tuning raises), so capability 5 is feasible here. The objection is the format id, not the CPU mode.

Recommendation, updated for the current body: capabilities 1–4 and 6 are ready to build; **5** belongs in its own issue.

---

<sub>Reviewer[design] · reviewed `dd2f55e` (tutorials/ + scripts/) · issue body at `11:38:13Z`, corrected against `11:44:02Z` · Claude Code / claude-opus-5 · provenance, not sign-off</sub>

**Comment by kurtvalcorza** [comment]
Thank you for the thorough design review and correction note. The RFC specification in the issue description has been updated to v3, adopting your recommendations in full:

1. **Dependency Pinning (N5):** Confirmed `"lightgbm>=4.0,<4.8"` will be added as a separate quoted token on the existing `%pip install` line, leaving the CI-required `'tabicl[finetune]==2.1.1'` literal untouched.
2. **Artifact Format ID & Packaging Boundary (N2 & Recommendation):** We adopt the recommendation to **exclude artifact bundling from this issue**. The exported bundle (`tabicl-regressor-artifact.zip`) and `ARTIFACT_FORMAT = 'tabicl-dimer-regressor-v1'` remain 100% standard and untouched, preserving contract compatibility with `tabiclv2_regressor_artifact_inference_colab.ipynb` and the container backend. Dual-mode demonstration is strictly an in-memory latency comparison; artifact-level tree packaging will be tracked in a separate issue.
3. **Point Predictions Only (N7):** Confirmed all comparisons and blending are strictly scoped to point predictions (one scalar per row), matching TabICLv2's existing API.
4. **Resolution Reporting & Mixed Evidence (N4):** The notebook will display the holdout one-row resolution directly beside reported metric deltas to contextualize small score movements, and will print a mixed-evidence warning if holdout and independent test metrics diverge.
5. **Reproducible Latency Benchmarks (N6):** Benchmark tables will print the resolved execution device (GPU/CPU) per row.

The issue is updated and ready for implementation.
